File: solutions/zig/queen-attack/1/queen_attack.zig
pub const QueenError = error{
    InitializationFailure,
};

pub const Queen = struct {
    row: i8,
    col: i8,

    pub fn init(row: i8, col: i8) QueenError!Queen {
        if (row < 0 or row > 7 or col < 0 or col > 7) {
            return QueenError.InitializationFailure;
        }
        return Queen{
            .row = row,
            .col = col,
        };
    }

    pub fn canAttack(self: Queen, other: Queen) QueenError!bool {
        return self.col == other.col or
            self.row == other.row or
            @abs(self.row - other.row) == @abs(self.col - other.col);
    }
};

File: solutions/lua/queen-attack/1/queen-attack.lua
return function(pos)
  assert(pos.row >= 0 and pos.row <= 7 and pos.column >= 0 and pos.column <= 7)
  return {
    pos = pos,
    can_attack = function(other)
      return pos.column == other.pos.column or
          pos.row == other.pos.row or
          math.abs(pos.row - other.pos.row) == math.abs(pos.column - other.pos.column)
    end
  }
end
